src/utilities/mta/mod.rs

```diff
@@ -41,79 +45,90 @@ pub struct MessageB {
+        (m_a, randomness)
     }
 
     pub fn a_with_predefined_randomness(
         a: &FE,
         alice_ek: &EncryptionKey,
         randomness: &BigInt,
+        dlog_statements: &[DLogStatement],
     ) -> Self {
         let c_a = Paillier::encrypt_with_chosen_randomness(
             alice_ek,
             RawPlaintext::from(a.to_big_int()),
             &Randomness::from(randomness.clone()),
-        );
+        )
+        .0
+        .clone()
+        .into_owned();
+        let alice_range_proofs = dlog_statements
+            .into_iter()
+            .map(|dlog_statement| {
+                AliceProof::generate(&a.to_big_int(), &c_a, alice_ek, dlog_statement, &randomness)
+            })
+            .collect::<Vec<AliceProof>>();
 
         Self {
-            c: c_a.0.clone().into_owned(),
+            c: c_a,
+            range_proofs: alice_range_proofs,
         }
     }
 }
 
 impl MessageB {
-    pub fn b(b: &FE, alice_ek: &EncryptionKey, c_a: MessageA) -> (Self, FE, BigInt, BigInt) {
+    pub fn b(
+        b: &FE,
+        alice_ek: &EncryptionKey,
+        m_a: MessageA,
+        dlog_statements: &[DLogStatement],
+    ) -> Result<(Self, FE, BigInt, BigInt), Error> {
         let beta_tag = BigInt::sample_below(&alice_ek.n);
-        let beta_tag_fe: FE = ECScalar::from(&beta_tag);
         let randomness = BigInt::sample_below(&alice_ek.n);
-        let c_beta_tag = Paillier::encrypt_with_chosen_randomness(
+        let (m_b, beta) = MessageB::b_with_predefined_randomness(
+            b,
             alice_ek,
-            RawPlaintext::from(beta_tag.clone()),
-            &Randomness::from(randomness.clone()),
-        );
+            m_a,
+            &randomness,
+            &beta_tag,
+            dlog_statements,
+        )?;
 
-        let b_bn = b.to_big_int();
-        let b_c_a = Paillier::mul(
-            alice_ek,
-            RawCiphertext::from(c_a.c),
-            RawPlaintext::from(b_bn),
-        );
-        let c_b = Paillier::add(alice_ek, b_c_a, c_beta_tag);
-        let beta = FE::zero().sub(&beta_tag_fe.get_element());
-        let dlog_proof_b = DLogProof::prove(b);
-        let dlog_proof_beta_tag = DLogProof::prove(&beta_tag_fe);
-
-        (
-            Self {
-                c: c_b.0.clone().into_owned(),
-                b_proof: dlog_proof_b,
-                beta_tag_proof: dlog_proof_beta_tag,
-            },
-            beta,
-            randomness,
-            beta_tag,
-        )
+        Ok((m_b, beta, randomness, beta_tag))
     }
 
     pub fn b_with_predefined_randomness(
         b: &FE,
         alice_ek: &EncryptionKey,
-        c_a: MessageA,
+        m_a: MessageA,
         randomness: &BigInt,
         beta_tag: &BigInt,
-    ) -> (Self, FE) {
+        dlog_statements: &[DLogStatement],
+    ) -> Result<(Self, FE), Error> {
+        if m_a.range_proofs.len() != dlog_statements.len() {
```

Comment:
I think another check is needed for the exact Len  - it should be equal to t  . 
Unless we assume message delivery is guaranteed ?

Reply:
Dlog statements vector is formed by taking only those elements of `h1_h2_n_tilde_vec` that participate in signing. So I guess its length is guaranteed to be correct?
Secondly, I think there are use-cases where dlog_statements might be an empty slice. I used it this way to omit range proofs in old gg18 tests where there's no h1, h2, N_tilde generation at all (I don't think gg18 part is really used by anyone at this point, right? Otherwise we should add h1, h2, N_tilde generation to it's `Keys::create` method). Another example: in blame, we want to get ciphertexts to compare them to the ones we've seen, but don't want to prove or verify any range proofs. Thus, I passed empty slices as `dlog_statements` when checking blame.